ff_derive/src/prime_field.rs

```diff
@@ -0,0 +1,830 @@
+                    let mut m = c0.wrapping_mul(INV);;
+                    c = ::fff::mac_with_carry_simple(c0, m, MODULUS.0[0], &mut c2);
+                });
+                for j in 1..limbs {
+                    let temp1 = get_temp(j);
+                    gen.extend(quote! {
+                        c0 = ::fff::mac_with_carry_d(c1, v, (#b.0).0[#j], #temp1, &mut c1);
+                    });
+                    let temp2 = get_temp(j - 1);
+                    let temp3 = get_temp(limbs - 1);
+                    if j == limbs - 1 {
+                        gen.extend(quote! {
+                            #temp2 = ::fff::mac_with_carry_e(c0 , m, MODULUS.0[#j], c2, c1,  &mut #temp3);
+                        });
+                    } else {
+                        gen.extend(quote! {
+                            #temp2 = ::fff::mac_with_carry_d(c2, m, MODULUS.0[#j], c0, &mut c2);
+                        });
+                    }
+                }
+            } else {
+                let temp3 = get_temp(0);
+                gen.extend(quote! {
+                    c0 = ::fff::mac_with_carry_simple(#temp3, v, (#b.0).0[0], &mut c1);
+                    let mut m = c0.wrapping_mul(INV);;
+                    c = ::fff::mac_with_carry_simple(c0, m, MODULUS.0[0], &mut c2);
+                });
+                for j in 1..limbs {
+                    let temp4 = get_temp(j);
+                    gen.extend(quote! {
+                        c0 = ::fff::mac_with_carry_d(c1, v, (#b.0).0[#j], #temp4, &mut c1);
+                    });
+                    let temp1 = get_temp(j - 1);
+                    if j == limbs - 1 {
+                        let temp0 = get_temp(limbs - 1);
+                        gen.extend(quote! {
+                            #temp1 = ::fff::mac_with_carry_e(c0, m, MODULUS.0[#j], c2,  c1, &mut #temp0);
+                        });
+                    } else {
+                        gen.extend(quote! {
+                            #temp1 = ::fff::mac_with_carry_d(c2, m, MODULUS.0[#j], c0, &mut c2);
+                        });
+                    }
+                }
+            }
+        }
+
+        for i in 0..limbs {
+            let temp = get_temp(i);
+
+            gen.extend(quote! {
+                (self.0).0[#i] = #temp;
+            });
+        }
+
+        gen.extend(quote! {
+            self.reduce();
+        });
+        gen
+    }
+
+    /*
+        Explicitly ported implementation from https://github.com/ConsenSys/goff/blob/master/internal/templates/element/mul_cios.go.
+        It is suggested that current implementation is pretty much the same in
+        terms of performance.
+     */
+    fn mul_impl_cios(
+        a: proc_macro2::TokenStream,
+        b: proc_macro2::TokenStream,
+        limbs: usize,
+    ) -> proc_macro2::TokenStream {
+        let mut gen = proc_macro2::TokenStream::new();
+
+        for i in 0..limbs * 2 + 1 {
+            let tempi = get_temp(i);
```

Comment:
Something seems wrong with the `tempi` allocations. You allocate `limbs * 2 + 1` temp variables, but `get_temp` is only ever called with values greater or equal to `limbs`

Reply:
Agreed. This can be reduced. Done.